File: Travel Package management C++/PassengerTest.cpp
#include "Passenger.h"
#include "StandardPassenger.h"
#include "GoldPassenger.h"
#include "PremiumPassenger.h"
#include "Activity.h"
#include <gtest/gtest.h>
#include <memory>


TEST(StandardPassengerTest, SignUpForActivity) {
    auto passenger = std::make_shared<StandardPassenger>("Amaan", 1, 200);
    auto destination = std::make_shared<Destination>("Bangalore");
    auto activity1 = std::make_shared<Activity>("Biking", "Mountain biking", 50, 5);
    auto activity2 = std::make_shared<Activity>("Safari", "Jungle Safari", 500, 5);

    destination->addActivity(activity1);
    destination->addActivity(activity2);

    EXPECT_TRUE(passenger->signUpForActivity(destination, activity1));
    EXPECT_FALSE(passenger->signUpForActivity(destination, activity2));
}

TEST(PremiumPassengerTest, SignUpForActivity) {
    auto passenger = std::make_shared<PremiumPassenger>("Amaan", 1, 200);
    auto destination = std::make_shared<Destination>("Bangalore");
    auto activity1 = std::make_shared<Activity>("Biking", "Mountain biking", 50, 5);
    auto activity2 = std::make_shared<Activity>("Safari", "Jungle Safari", 500, 5);

    destination->addActivity(activity1);
    destination->addActivity(activity2);

    EXPECT_TRUE(passenger->signUpForActivity(destination, activity1));
    EXPECT_TRUE(passenger->signUpForActivity(destination, activity2));
}

TEST(GoldPassengerTest, ActivityCost) {
    auto passenger = std::make_shared<GoldPassenger>("Amaan", 1, 200);
    auto destination = std::make_shared<Destination>("Bangalore");
    auto activity1 = std::make_shared<Activity>("Biking", "Mountain biking", 50, 5);

    destination->addActivity(activity1);

    passenger->signUpForActivity(destination, activity1);

    EXPECT_EQ(passenger->calculateActivityCost(activity1), 45);
}

TEST(PremiumPassengerTest, ActivityCost) {
    auto passenger = std::make_shared<PremiumPassenger>("Amaan", 1, 200);
    auto destination = std::make_shared<Destination>("Bangalore");
    auto activity1 = std::make_shared<Activity>("Biking", "Mountain biking", 50, 5);

    destination->addActivity(activity1);

    passenger->signUpForActivity(destination, activity1);

    EXPECT_EQ(passenger->calculateActivityCost(activity1), 0);
}

TEST(StandardPassengerTest, ActivityCost) {
    auto passenger = std::make_shared<StandardPassenger>("Amaan", 1, 200);
    auto destination = std::make_shared<Destination>("Bangalore");
    auto activity1 = std::make_shared<Activity>("Biking", "Mountain biking", 50, 5);

    destination->addActivity(activity1);

    passenger->signUpForActivity(destination, activity1);

    EXPECT_EQ(passenger->calculateActivityCost(activity1), 50);
}

File: Travel Package Management Java/src/com/example/travel/PassengerTest.java
package com.example.travel;

import org.junit.Test;
import static org.junit.Assert.*;

public class PassengerTest {
    
    @Test
    public void testSignUpForActivity_StandardPassenger() {
        Destination bangalore = new Destination("Bangalore");
        Activity biking = new Activity("Biking", "Mountain biking", 50, 5);
        bangalore.addActivity(biking);

        StandardPassenger amaan = new StandardPassenger("Amaan", 1, 200);
        assertTrue(amaan.signUpForActivity(bangalore, biking));
        assertFalse(amaan.signUpForActivity(bangalore, biking)); // Already signed up
    }

    @Test
    public void testSignUpForActivity_GoldPassenger() {
        Destination bangalore = new Destination("Bangalore");
        Activity biking = new Activity("Biking", "Mountain biking", 50, 5);
        bangalore.addActivity(biking);

        GoldPassenger amaan = new GoldPassenger("Amaan", 1, 200);
        assertTrue(amaan.signUpForActivity(bangalore, biking)); // GoldPassenger signing up
        assertTrue(biking.isSignedUp(amaan)); // Confirm signup
    }

    @Test
    public void testActivityCost_GoldPassenger() {
        Activity biking = new Activity("Biking", "Mountain biking", 50, 5);
        GoldPassenger amaan = new GoldPassenger("Amaan", 1, 200);

        double expectedCost = 45.0; // Expecting a 10% discount
        assertEquals(expectedCost, amaan.calculateActivityCost(biking), 0.01);
    }

    @Test
    public void testSignUpForActivity_PremiumPassenger() {
        Destination bangalore = new Destination("Bangalore");
        Activity biking = new Activity("Biking", "Mountain biking", 50, 5);
        bangalore.addActivity(biking);

        PremiumPassenger amaan = new PremiumPassenger("Amaan", 1, 200);
        assertTrue(amaan.signUpForActivity(bangalore, biking)); // PremiumPassenger should sign up for free
        assertTrue(biking.isSignedUp(amaan)); // Confirm signup
    }

    @Test
    public void testActivityCost_PremiumPassenger() {
        Activity biking = new Activity("Biking", "Mountain biking", 50, 5);
        PremiumPassenger amaan = new PremiumPassenger("Amaan", 1, 200);

        double expectedCost = 0.0; // PremiumPassenger attends for free
        assertEquals(expectedCost, amaan.calculateActivityCost(biking), 0.01);
    }

   
}
